core/system_metric/sys_metric_stat.go

```diff
@@ -15,50 +15,79 @@
 package system_metric
 
 import (
-	"math"
+	"os"
 	"sync"
 	"sync/atomic"
 	"time"
 
 	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/metrics"
 	"github.com/alibaba/sentinel-golang/util"
-	"github.com/shirou/gopsutil/v3/cpu"
 	"github.com/shirou/gopsutil/v3/load"
+	"github.com/shirou/gopsutil/v3/mem"
+	"github.com/shirou/gopsutil/v3/process"
 )
 
 const (
-	NotRetrievedValue float64 = -1.0
+	NotRetrievedLoadValue     float64 = -1.0
+	NotRetrievedCpuUsageValue float64 = -1.0
+	NotRetrievedMemoryValue   int64   = -1
 )
 
 var (
-	currentLoad     atomic.Value
-	currentCpuUsage atomic.Value
+	currentLoad        atomic.Value
+	currentCpuUsage    atomic.Value
+	currentMemoryUsage atomic.Value
 
-	prevCpuStat *cpu.TimesStat
-	initOnce    sync.Once
+	loadStatCollectorOnce   sync.Once
+	memoryStatCollectorOnce sync.Once
+	cpuStatCollectorOnce    sync.Once
+
+	CurrentPID         = os.Getpid()
+	currentProcess     atomic.Value
+	currentProcessOnce sync.Once
+	TotalMemorySize    = getTotalMemorySize()
 
 	ssStopChan = make(chan struct{})
 )
 
 func init() {
-	currentLoad.Store(NotRetrievedValue)
-	currentCpuUsage.Store(NotRetrievedValue)
+	currentLoad.Store(NotRetrievedLoadValue)
+	currentCpuUsage.Store(NotRetrievedCpuUsageValue)
+	currentMemoryUsage.Store(NotRetrievedMemoryValue)
+
+	p, err := process.NewProcess(int32(CurrentPID))
+	if err != nil {
+		logging.Error(err, "Fail to new process when initializing system metric", "pid", CurrentPID)
+	}
+	currentProcessOnce.Do(func() {
+		currentProcess.Store(p)
```

Comment:
Maybe we should not set currentProcess when err != nil

Reply:
Exactly, I will update here